### PR TITLE
More testing for gRPC and make gRPC Caller and Service headers required

### PR DESCRIPTION
**Pull request by peter-edge**
- Adds integration testing for gRPC
- Adds a private unary interceptor option so that metadata can be intercepted in integration testing
- Make gRPC caller and service headers required
- Includes some refactoring of the examples

### CODE REVIEW

File: transport/x/grpc/options.go [hunk 1]
```diff
@@ -61,6 +65,13 @@ func (i *inboundOptions) getTracer() opentracing.Tracer {
 	return i.tracer
 }
 
+// TODO: this should cover the tracer interceptor too
+// grpc-go only allows one interceptor, so need to handle all cases
+// working on this with go-grpc-middleware
+func (i *inboundOptions) getUnaryInterceptor() grpc.UnaryServerInterceptor {
```
Comment (willhug) [inline transport/x/grpc/options.go]: I thought this diff was only for tests... what's this stuff?

File: transport/x/grpc/handler.go [hunk 2]
```diff
@@ -102,18 +102,9 @@ func (h *handler) getTransportRequest(ctx context.Context, decodeFunc func(inter
 	if err != nil {
 		return nil, err
 	}
-	if transportRequest.Caller == "" {
```
Comment (willhug) [inline transport/x/grpc/handler.go]: is this right?

File: transport/x/grpc/options.go (hunk 1 above)
Comment (peter-edge) [inline transport/x/grpc/options.go]: This is for testing only right now

File: transport/x/grpc/handler.go (hunk 2 above)
Comment (peter-edge) [inline transport/x/grpc/handler.go]: Yes

**Comment by peter-edge** [comment]
The interceptor stuff is used in the integration test and was needed for the metadata testing (which was the initial reason for this PR). Splitting this PR up at this point is going to be difficult, and I don't think would help much, is this really necessary?

**Pull request merged**